whippet deploy can optionally deploy public dirs

snim2 opened the pull request:
Before: any files in the `public/` directory inside the WordPress app were copied (recursively) into the app directory on deploy.

Now: by passing in the `-p (--public) <directory>` switch to 'whippet deploy' the `public/` directory can be deployed to the given destination directory.

Resolves: #196

## Testing locally

1. Move somewhere where you don't mind splatting a lot of files, e.g. `cd /tmp`
2. Clone any Whippetised repo, e.g. `git clone git@github.com:dxw/saluki-test-site.git`
3. Create deployment directories, e.g. `mkdir -r /tmp/whippet-deploy/shared/uploads`
4. Add a basic `wp-config.php` file to your `shared/` directory, e.g.:
```php
define( 'DB_NAME', 'test_tb' );
define( 'DB_USER', 'db_user' );
define( 'DB_PASSWORD', 'db_passwd' );
define( 'DB_HOST', 'localhost' );
```
5. Ensure that you have this branch of Whippet checked out
6. From the root of your cloned Wordpress repo, run some test deploys, e.g. ` /path/to/whippet/bin/whippet deploy -f /tmp/whippet-deploy -p /tmp/random-directory`
7. Check that `/tmp/random-directory` contains the files you can see in `public/`

----

- [ ] Includes tests (if new features are introduced)
- [x] Commit message includes link to the ticket/issue
- [x] Changelog updated
- [ ] PR open against [dxw's Homebrew Tap](https://github.com/dxw/homebrew-tap/) to point the Whippet formula to the new version number

snim2 commented on src/Modules/Deploy.php:
```diff
@@ -33,6 +34,14 @@ public function deploy($force, $keep)
             $this->check_and_create_dir($this->shared_dir);
+            // If this deploy is _not_ deploying public/ as a separate
+            // directory, we ignore any existing files in public/.
+            if ($public) {
```
@leedxw if you wanted this done with a symlink please let me know!

RobjS commented on docs/deploy.md:
````diff
@@ -34,3 +34,25 @@ Note also that Whippet, by default, will not deploy your application if the comm
+$ whippet deploy -p /var/local/myapp
+```
+
+will copy the contents of `public/` to `/var/local/public`, rather than copying
````
This doesn't quite describe what I'm seeing, which is that `whippet deploy -p /var/local/myapp` results in a public directory at `/var/local/myapp/public/`, as the app lives in `/var/local/myapp/current/` (symlinked to the appropriate folder in `/var/local/myapp/releases/`).

snim2 replied:
Hmm, this is what I'm seeing:

```
❯ ls
total 0
drwxr-xr-x@ 6 snim2  wheel   192B 10 Jun 21:34 releases/
lrwxr-xr-x@ 1 snim2  wheel    74B 10 Jun 21:34 current -> /private/tmp/whip-deploy/releases/4451fc0e70d9cb8952482405b50150a8c8ff21eb
drwxr-xr-x@ 3 snim2  wheel    96B 10 Jun 21:34 public/
drwxr-xr-x@ 4 snim2  wheel   128B 10 Jun 21:18 shared/
```

but the `public/` directory lives in the app repo too, so maybe the question is should we delete it there when we deploy with `-p`?

RobjS replied:
Yes, that matches what I'm seeing - but doesn't that mean that `whippet deploy -p /var/local/myapp` gives you a public folder at `/var/local/myapp/public`, not at `/var/local/public` as the docs claim (as all the folders listed there live inside `myapp/`)?

snim2 commented:
@leedxw and I had a chat about this, and:

1. The deployed directory should be configurable on the command line
1. We want to copy files, not symlink
1. We want to copy files, we do _not_ want to delete existing files

RobjS commented:
I'm getting a PHP Notice if I run `whippet deploy` without the `-p` option:
```bash
whippet-dev deploy ~/Projects/whippet-deploys/bikeshed
PHP Notice:  Undefined property: stdClass::$public in /Users/rob/projects/whippet/src/Whippet.php on line 53
Notice: Undefined property: stdClass::$public in /Users/rob/projects/whippet/src/Whippet.php on line 53
```

snim2 commented:
> I'm getting a PHP Notice if I run `whippet deploy` without the `-p` option:
> 
> ```shell
> whippet-dev deploy ~/Projects/whippet-deploys/bikeshed
> PHP Notice:  Undefined property: stdClass::$public in /Users/rob/projects/whippet/src/Whippet.php on line 53
> Notice: Undefined property: stdClass::$public in /Users/rob/projects/whippet/src/Whippet.php on line 53
> ```

Thanks, pushed a fix

RobjS commented:
Once I've done a first deploy to using the `-p` option, I get a fatal error next time I try to do a deploy with the same `-p` option, unless I remove that directory first, e.g. (following a successful run of the same command the first time):
```bash
 whippet-dev deploy ~/Projects/whippet-deploys/bikeshed -f -p ~/Projects/whippet-deploys/bikeshed/some-public-dir
PHP Fatal error:  Uncaught Exception: Directory exists but is not empty: /Users/rob/Projects/whippet-deploys/bikeshed/some-public-dir in /Users/rob/projects/whippet/src/Modules/Helpers/WhippetHelpers.php:15
Stack trace:
#0 /Users/rob/projects/whippet/src/Modules/Deploy.php(41): Dxw\Whippet\Modules\Deploy->check_and_create_dir('/Users/rob/Proj...', true)
#1 /Users/rob/projects/whippet/src/Whippet.php(57): Dxw\Whippet\Modules\Deploy->deploy(true, 3, '/Users/rob/Proj...')
#2 /Users/rob/projects/whippet/vendor/dxw/rubbishthorclone/rubbish_thor_clone.class.php(69): Dxw\Whippet\Whippet->deploy('/Users/rob/Proj...')
#3 /Users/rob/projects/whippet/bin/whippet(26): RubbishThorClone->start(Array)
#4 {main}
  thrown in /Users/rob/projects/whippet/src/Modules/Helpers/WhippetHelpers.php on line 15\
```

snim2 commented on src/Modules/Deploy.php:
```diff
@@ -33,6 +34,12 @@ public function deploy($force, $keep)
             $this->check_and_create_dir($this->deploy_dir);
             $this->check_and_create_dir($this->releases_dir);
             $this->check_and_create_dir($this->shared_dir);
+            // If this deploy is _not_ deploying public/ as a separate
```
This needs another look

snim2 commented:
Bringing this branch up to date (so, force-pushing)

snim2 commented:
@RobjS This is ready for re-review. I have:

* Squashed and brought this up to date with current `main` (you may have to nuke any local copies of this branch, sorry)
* Updated PR description
* Updated docs in `docs/deploy.md` to match the changes here

RobjS reviewed (approved):
This is all working for me now, and repeat deploys with or without the `-p` option are behaving as expected 👍
